crates/typst/src/eval/version.rs

```diff
@@ -0,0 +1,223 @@
+        }
+    };
+    ($($input:tt)*) => {
+        component_names!(@gen [] [0] [] $($input)*);
+    };
+}
+
+impl Version {
+    component_names!["major", "minor", "patch"];
+
+    pub fn new() -> Self {
+        Self(EcoVec::new())
+    }
+
+    fn get(&self, index: usize) -> Option<VersionComponent> {
+        self.0.get(index).copied()
+    }
+
+    /// Get a component of a version.
+    ///
+    /// Always non-negative. Returns `0` if the version isn't specified to the necessary length.
+    pub fn at(&self, index: i64) -> StrResult<i64> {
+        if index < 0 {
+            bail!("version component index out of bounds ({index})");
+        }
+        Ok(usize::try_from(index)
+            .ok()
+            .and_then(|i| self.get(i))
+            .unwrap_or_default()
+            .get())
+    }
+
+    /// Get a named component of a version.
+    ///
+    /// Always non-negative. Returns `0` if the version isn't specified to the necessary length.
+    pub fn component(&self, name: &str) -> StrResult<i64> {
+        match Self::resolve_component_name(name) {
+            Some(i) => Ok(self.get(i).unwrap_or_default().get()),
+            None => bail!("unknown version component"),
+        }
+    }
+
+    /// Get an iterator of the values of all named version components for a version.
+    pub fn named_components(&self) -> impl Iterator<Item = (&'static str, i64)> + '_ {
+        Self::COMPONENT_NAMES
+            .into_iter()
+            .enumerate()
+            .map(|(i, name)| (name, self.get(i).unwrap_or_default().get()))
+    }
+
+    /// Convert a version into an array
+    pub fn into_array(self) -> Array {
+        self.0.into_iter().map(|i| Value::Int(i.get())).collect()
+    }
+
+    pub fn push(&mut self, component: VersionComponent) {
+        self.0.push(component);
+    }
+}
+
+impl FromIterator<VersionComponent> for Version {
+    fn from_iter<T: IntoIterator<Item = VersionComponent>>(iter: T) -> Self {
+        Self(EcoVec::from_iter(iter))
+    }
+}
+
+impl Ord for Version {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let max_len = self.0.len().max(other.0.len());
+        let tail = repeat(&VersionComponent::ZERO);
+
+        let self_iter = self.0.iter().chain(tail.clone());
+        let other_iter = other.0.iter().chain(tail);
+
+        for (l, r) in self_iter.zip(other_iter).take(max_len) {
+            match l.cmp(r) {
+                Ordering::Equal => (),
+                ord => return ord,
+            }
+        }
+        Ordering::Equal
+    }
+}
+impl PartialOrd for Version {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl Eq for Version {}
+impl PartialEq for Version {
+    fn eq(&self, other: &Self) -> bool {
+        matches!(self.cmp(other), Ordering::Equal)
+    }
+}
+
+impl Hash for Version {
+    // Don't hash any explicitly suffixed zeros so that equal but non-identical versions (e.g. 1.2 and 1.2.0) hash as the same version
```

Comment:
I'm not sure this makes sense. Do `1.2` and `1.2.0` display differently? If so, they **must** hash differently. And if not, why even store the zeros in the first place?

Reply:
Well, they display differently, but they compare as equal. Maybe the displaying should be changed, then (and suffix zeros would just not be stored), because the comparing as equal part is pretty important IMO.

Reply:
Stripping would be kind of bad though because semver requires exactly three components. I agree that comparing equal is good, but the really must hash differently because otherwise it'd break incremental compilation. That's even more important than std's statements about Hash and Eq.

Reply:
So you're saying the Eq and Hash should be different in this case?
Or maybe we should let Rust-Eq see them as different but add another method that's used for Typst-Eq?

Reply:
Just make it Hash and Eq different. It's already the case for a bunch of types in Typst.

Reply:
Okay, that's good then.